src/Lecture2.hs

```diff
@@ -292,4 +437,25 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding = toExpr . sumConst . toTerms
+  where
+    toTerms :: Expr -> [Expr]
+    toTerms expr = case expr of
+      l@(Lit _) -> [l]
+      v@(Var _) -> [v]
+      Add e1 e2 -> toTerms e1 ++ toTerms e2
+
+    sumConst :: [Expr] -> (Int, [Expr])
+    sumConst = go 0 []
+      where
+        go val exprs [] = (val, exprs)
+        go val exprs (x : xs) = case x of
+          (Lit l) -> go (val + l) exprs xs
+          e -> go val (e : exprs) xs
```

Comment:
Great implementation!

It looks nice and it's quite okay to leave as it is for a simple data type like `Expr`. However, when the type becomes bigger, you may want to reflect only valid states in the type.

For example, the `toTerms` function says that it returns a list of `Expr` values but it doesn't actually return the `Add` constructor. A simple way to do this here would be to change the resulting type from `[Expr]` to `[Either Int String]`.

Similarly, the resulting type of `sumConst` could be changed from `(Int, [Expr])` to `(Int, [String])` because you don't return the list of any expressions but the list of variable names solely.

And in the end, you can have a helper function to combine all variable names with `Add`:

```haskell
addVars :: String -> [String] -> Expr
addVars v vs = foldr Add (Var v) (map Var vs)
```

Reply:
Сorrected.